Support for UDP packet mode

helgeolav opened the pull request:
As discussed in #146 i have made some changes to the code to handle UDP packets in a packet-oriented way instead of the stream-oriented way that the socket input handles UDP traffic. By setting a new configuration option "packetmode" to true each packet is processed separately and sent in for further processing. A field "host_ip" is also added that is the source IP/port for the packet.

tsaikd commented on input/socket/inputsocket.go:
```diff
@@ -121,7 +123,11 @@ func (i *InputConfig) Start(ctx context.Context, msgChan chan<- logevent.LogEven
 			return err
 		}
-		return i.handleUDP(ctx, conn, msgChan)
+		if i.PacketMode == false {
+			return i.handleUDP(ctx, conn, msgChan)
+		} else {
+			return i.handleUDPpacketMode(ctx, conn, msgChan)
+		}
```
```suggestion
		if i.PacketMode {
			return i.handleUDPpacketMode(ctx, conn, msgChan)
		} else {
			return i.handleUDP(ctx, conn, msgChan)
		}
```

tsaikd commented on input/socket/inputsocket.go:
```diff
@@ -31,6 +31,8 @@ type InputConfig struct {
 	ReusePort  bool   `json:"reuseport"`
 	BufferSize int    `json:"buffer_size"`
+	// packetmode is only valid for UDP sessions and handles each packet as a message on its own
+	PacketMode bool `json:"packetmode"`
```
```suggestion
	PacketMode bool `json:"packet_mode"`
```

tsaikd commented on input/socket/inputsocket.go:
```diff
@@ -152,6 +158,54 @@ func (i *InputConfig) Start(ctx context.Context, msgChan chan<- logevent.LogEven
+				}
+			}
+			// handle error
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				return err
+			}
+		}
+		return nil
```
```suggestion
			switch err {
			case nil:
				// continue processing
			case io.EOF:
				return nil
			default:
				return err
			}
```

tsaikd commented on input/socket/inputsocket.go:
```diff
@@ -152,6 +158,54 @@ func (i *InputConfig) Start(ctx context.Context, msgChan chan<- logevent.LogEven
+				} else {
+					extras := make(map[string]interface{})
+					extras["host_ip"] = addr.String()
+					_, _ = i.Codec.Decode(ctx, b[:n], extras, []string{}, msgChan)
```
ignore all decode errors?

helgeolav replied:
I based my code on the existing code. All other code paths also ignore the result from Decode. I only added _, _ to be explicit about me ignoring the result.

If I should handle this I am a bit unsure how to handle the result. Please advice. I also should handle the error in the other code paths as well?

As Decode returns (ok bool, err error) I have four possible results:

```text
ok err
======
fa nil message not sent, no error to explain why
fa err message not sent with error
tr nil everything is ok, normal behaviour
tr err message sent, with error
```
I could write some code to send a log message in case the message was not sent or there is an error?

aantono replied:
The method signature is more nuanced.
The real failure is ONLY when `err` is returned.  That's a clear indication that the processing of data input has failed, either with parsing or with some other, codec-specific, data processing/decoding.
If `ok` is false, that only means that a new `Event` struct was not sent to the `msgChan`.  This could happen IFF a particular Codec implementation does not do any `Event` decoding per-say, which results in a new `Event`, but rather extracts some of the data and either does some "pre-processing" or adds attributes to the `Context` object.  (Neither of those are common/typical cases, but theoretically they both could occur)

In your particular case, I would only check for `err != nil` as a check.

tsaikd commented on input/socket/inputsocket.go:
```diff
@@ -152,6 +158,54 @@ func (i *InputConfig) Start(ctx context.Context, msgChan chan<- logevent.LogEven
+					i.BufferSize += bufIncSize
+					b = make([]byte, i.BufferSize)
+				} else {
+					extras := make(map[string]interface{})
+					extras["host_ip"] = addr.String()
```
```suggestion
					extras := map[string]interface{}{
						"host_ip": addr.String(),
					}
```

helgeolav commented:
I have updated my fork with all the changes that have been discussed up to now. I also added error handling for Codec in parse() so the behaviour is similar however this input is used.

tsaikd commented on input/socket/inputsocket.go:
```diff
@@ -152,6 +158,61 @@ func (i *InputConfig) Start(ctx context.Context, msgChan chan<- logevent.LogEven
+				return err
+			}
+		}
+		return nil
```
```suggestion
```

unreachable code

helgeolav replied:
I have removed line 210 now. Please accept change.

tsaikd commented:
Sorry for replying lately, please rearrange your commits before merging.

helgeolav commented:
Closing, change was implemented at #152.
